### DISCUSSION TITLE
Add `toPromise` for `useQuery()`'s `executeQuery`

### DISCUSSION THREAD

**Discussion by nandorojo** [urql-graphql/urql#2145] [Q&A] [answered]
A common UX pattern I like from React Query is, click a link, prefetch that page, and then navigate.

I want to be able to "prefetch" something with urql, and then navigate once it's done.

I've tried this:

```tsx
const LinkToArtist = () => {
  const [query, execute] = useQuery({
    query: `...`,
    pause: true
  })

  const onPress = async () => {
    await execute() // this isn't a promise 😕
    Router.push('/artist')
  }

  // ...
}
```

I could, alternatively, just `useClient().query().toPromise()`, but then I don't get the nice loading states (and types from my graphql codegen).

Is there a way to get the `execute` function returned by `useQuery` to return a promise, just like `useMutation` does?

Thanks!

**Comment by kitten** [comment, marked as answer]
This sounds a little like a Suspense usecase to me, doesn't it? Instead of relying on this manually it seems to me like suspense attempts to solve this problem at the root 

I've previously tried to implement this and it rarely, imho, is a good pattern when it's implemented at the navigation site. That's because you have `n` places that could navigate to a page but only `1` target that these loads end up in.

So, any pattern that allows this is certainly one that should in some shape or form rely on the target of the navigation, not the source.

That said, the reason why this doesn't exist in the API is simple. If an API addition in urql is already covered by something else, then we've historically been very carefully leaning on the side of not adding it. There's almost never multiple ways to implement a single pattern (maybe with the exception of pagination 😛 but that's so dependent on the schema and UI anyway)

On another note, managing this many hooks to be paused and loaded for an upcoming navigation seems like a lot of trouble brewing and ready to backfire 🤔

Edit: also, to add some colour to this. There's a bit of API design going on in the background here, and encouragement of certain patterns.

Queries are reactively observable. They only have a single owner and can be activated from a bindibg's site. They're very intentionally not imperative. Even if you use the promise conversion you're creating an observable source and are then guaranteeing to only observe the result in a specific place (at call site, unless you hand the result off)

This isn't the same for bindings. Whenever you use a query, there's actually a guarantee around how that query is managed and it can be cancelled throughout. Promises break this pattern — especially in React. It would be impossible for us to implement a reactive binding to a query _and_ allow you to observe the result in a promise.

Mutations are different because they're always fired imperatively and never cancelled

**Reply by nandorojo** [reply]
> This sounds a little like a Suspense usecase to me, doesn't it? Instead of relying on this manually it seems to me like suspense attempts to solve this problem at the root

Certainly. But I’m not exactly sure how to achieve the same in suspense quite yet.

Fair points. I suppose I’m mostly constrained by the fact that my codegen doesn’t export a nice `executeQuery` wrapper. I’m so spoiled by generated, typesafe code, that I am averse to directly importing from urql and typing it myself. But I’ll stick to that. Thanks for the insight.

**Reply by nandorojo** [reply]
> I've previously tried to implement this and it rarely, imho, is a good pattern when it's implemented at the navigation site. That's because you have n places that could navigate to a page but only 1 target that these loads end up in.

> So, any pattern that allows this is certainly one that should in some shape or form rely on the target of the navigation, not the source.

I think I get what you’re saying here…do you mean that the navigation system should determine whether or not to navigate? If so, I agree it’s the right abstraction, but I’m not sure if the APIs have caught up yet. So I’m just trying to retro-fit this suspenseful pattern for now.

**Reply by kitten** [reply]
> my codegen doesn’t export a nice executeQuery wrapper

Do you have some information on what codegen you're using? We're still waiting for it all to fall into place before recommending anything in the docs, but besides that we'd always recommend `TypedDocumentNode` generation, e.g. with GraphQL Code Generator, as it provides type safety with any and every API without any wrapper 👀  https://www.graphql-code-generator.com/plugins/typed-document-node
